metis-dereference/metis-dereference-service/src/main/java/eu/europeana/metis/dereference/service/MongoDereferenceService.java

```diff
@@ -171,74 +172,125 @@ private void extractBroaderResources(EnrichmentBase resource, Set<String> destin
-      vocabulary = Optional.ofNullable(cachedEntity.getVocabularyId()).map(vocabularyDao::get)
-              .orElse(null);
+      transformedEntityXml = cachedEntity.getXml();
+      entityVocabulary = Optional.ofNullable(cachedEntity.getVocabularyId()).map(vocabularyDao::get)
+          .orElse(null);
     }
 
-    // If not in the cache, or no vocabulary (ID) is known, we need to resolve the resource.
-    if (entityString == null || vocabulary == null) {
-      final VocabularyCandidates candidates =
-          VocabularyCandidates.findVocabulariesForUrl(resourceId, vocabularyDao::getByUriSearch);
-      final Pair<String, Vocabulary> transformedEntity =
-          retrieveTransformedEntity(resourceId, candidates);
-      if (transformedEntity != null) {
-        entityString = transformedEntity.getLeft();
-        vocabulary = transformedEntity.getRight();
-        ProcessedEntity entityToCache = (cachedEntity == null) ? new ProcessedEntity() : cachedEntity;
-        entityToCache.setXml(entityString);
+    final Pair<String, Vocabulary> entityVocabularyPair = computeEntityAndVocabulary(resourceId,
+        cachedEntity, transformedEntityXml, entityVocabulary);
+
+    // Parse the entity.
+    final Pair<EnrichmentBase, Vocabulary> enrichmentBaseVocabularyPair;
+    if (entityVocabularyPair == null) {
+      enrichmentBaseVocabularyPair = null;
+    } else {
+      enrichmentBaseVocabularyPair = convertToEnrichmentBaseVocabularyPair(
+          entityVocabularyPair.getLeft(), entityVocabularyPair.getRight());
+    }
+    return enrichmentBaseVocabularyPair;
+  }
+
+  /**
+   * Based on the values of {@code transformedEntityXml} and {@code entityVocabulary} compute the
+   * pair with the transformed entity and vocabulary.
+   * <p>The possible outcomes are:
+   * <ul>
+   *   <li>If the {@code entityVocabulary} is not null then simply return a pair with the
+   *   provided values. We accept that the returned entity can be null, as a result of a
+   *   previous transformation.</li>
+   *   <li>If {@code entityVocabulary} is null then retrieve the original entity and transform it
+   *   based on the vocabulary candidates from the database, and return that pair</li>
+   * </ul>
+   * </p>
+   *
+   * @param resourceId the url of the provider entity
+   * @param cachedEntity the cached entity object
+   * @param transformedEntityXml the previously transformed provider entity
+   * @param entityVocabulary the previously used vocabulary for the transformation
+   * @return a pair with the computed values
+   * @throws URISyntaxException if the resource identifier url is invalid
+   * @throws TransformerException if an exception occurred during transformation of the original
+   * entity
+   */
+  private Pair<String, Vocabulary> computeEntityAndVocabulary(String resourceId,
+      ProcessedEntity cachedEntity, String transformedEntityXml, Vocabulary entityVocabulary)
+      throws URISyntaxException, TransformerException {
```

Comment:
I guess you extracted this code into a separate method due to concerns about complexity? I think you could perhaps split this up differently, though. You're passing too much information (the transformedEntityXml is also in the cachedEntity) and it does more than the name suggests (it also saves to the cache). Perhaps a rearrangement of the code makes more sense here?